libraries/extensions/ros2-bridge/msg-gen/src/parser/package.rs

```diff
@@ -1,94 +1,106 @@
-            .to_string();
-        // Hack
-        if file_name == "libstatistics_collector" {
-            continue;
-        }
+    // Return empty vector if root_dir is empty
+    if root_dir.as_ref() == Path::new("") {
+        let empty_vec: Vec<Package> = vec![];
+        warn!("AMENT_PREFIX_PATH pointed to ''");
+        return Ok(empty_vec);
+    }
 
-        let mut package = Package::new(file_name.clone());
-        for line in BufReader::new(File::open(path)?).lines() {
-            match parse_line(&line?) {
-                Some(("msg", v)) => {
-                    let msg = parse_message_file(
-                        &file_name,
-                        root_dir
-                            .as_ref()
-                            .join(format!("share/{}/msg/{}.msg", file_name, v)),
-                    )?;
-                    package.messages.push(msg);
-                }
-                Some(("srv", v)) => {
-                    let srv = parse_service_file(
-                        &file_name,
-                        root_dir
-                            .as_ref()
-                            .join(format!("share/{}/srv/{}.srv", file_name, v)),
-                    )?;
-                    package.services.push(srv);
+    for ros_format in ros_formats {
+        let pattern = root_dir.as_ref().to_string_lossy().to_string()
+            + "/**/"
+            + ros_format
+            + "/*."
+            + ros_format;
+        let mut visited_files = vec![];
+        for entry in glob(&pattern).expect("Failed to read glob pattern") {
+            match entry {
+                Ok(path) => {
+                    let file_name = path
+                        .clone()
+                        .file_name()
+                        .unwrap()
+                        .to_str()
+                        .unwrap()
+                        .to_string();
+
+                    let package = path
+                        .parent()
+                        .context("Should have a msg folder")?
+                        .parent()
+                        .context("should have a package folder")?
+                        .file_name()
+                        .context("folder name should exist")?
+                        .to_string_lossy()
+                        .to_string();
+
+                    // Hack
+                    if file_name == "libstatistics_collector" {
+                        continue;
+                    } else if visited_files.contains(&(package.clone(), file_name.clone())) {
+                        warn!(
+                        "found two versions of package: {:?}, message: {:?}. will skip the one in: {:#?}",
+                        package, file_name, path
+                    );
+                        continue;
+                    } else {
+                        visited_files.push((package.clone(), file_name.clone()));
+                    }
+
+                    match map.get_mut(&package) {
```

Comment:
We could use the [entry API](https://doc.rust-lang.org/stable/std/collections/struct.HashMap.html#method.entry) here to avoid the duplication:

```rust
let p = map.entry(package.clone).or_insert_with(|| Package::new(package.clone()));
```

Reply:
Indeed. Going to make the change.